### PR TITLE
modify discretized-ssm.ipynb to plot every time steps for the PF example

### PR DESCRIPTION
**Pull request by benlau6**
## Description

Modified the notebook to plot every time step for the particle filtering example, including true x(k), observed y(k), filtered mean E[x(k) | y(1:k)] and variance +- sqrt{Var[x(k)|y(1:k)} and smoothed mean E[x(k) | y(1:T)] and variance +- sqrt{Var[x(k)|y(1:T)}, in a line plot. Also plotted the full posterior p(x(k) | y(1:k)) and p(x(k)|y(1:T)) as a heatmap to capture multi-modality.

### Colab link

https://colab.research.google.com/github/benlau6/probml-notebooks/blob/discretized-ssm-patch/notebooks/discretized_ssm.ipynb

### Related PR
[Resolves #692](https://github.com/probml/pyprobml/pull/700) 

### Related Issue 

[exact 1d Bayesian filtering](https://github.com/probml/pyprobml/issues/692)

### Figures

[gist for line plot and heatmap](https://gist.github.com/benlau6/50039bab07a868ccd9cd87cec823e537)
## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

## Potential problems/Important remarks

@murphyk My code is ready as requested, please note that I replaced for loops with jax.lax.scan and also made a few changes as requested by gerdm at [pyprobml PR #700](https://github.com/probml/pyprobml/pull/700)

### CODE REVIEW

**Comment by benlau6** [comment]
@murphyk Updated. Please note that I marked red 'X's as ground truth states on heatmap. I also refactored the code a bit by correcting indentations, rearranging some code blocks, and removing some duplicated code blocks, functions, and line breaks.

**Comment by murphyk** [comment]
Thanks, the new visualizations look good. (Although in the line plot I would label the x axis as 'time' rather than then meaningless 'k' index.)

I am trying to find an example that will look good in the book, and which will make intuitive sense to the reader. The first and third example are rather complicated, but the second one is perfect: a Gaussian random walk in which some of the observations are outliers. This could in principle also be a good test bed for evaluating Kalman smoothing variants that can handle robust noise models.   However it seems there may be a bug in your code, since the performance of the smoother is worse than the filter, which seems wrong. This can be seen in your 2 plots:

Screen Shot 2022-04-16 at 10 59 02 AM


Screen Shot 2022-04-16 at 10 58 44 AM

Also when you plot a single time slice (like in fig 2 of the paper) you get this

Screen Shot 2022-04-16 at 10 59 41 AM

The smoothed distribution seems close to unimodal (as is to be expected, since there is less posterior uncertainty), but it is centered at the wrong point. Maybe the bug is in mapping from the discrete states to the correspond continuous values?

One way to debug this would be to replace the student likelihood with a Gaussian likelihood and then compare to the optimal estimate computed using my Kalman filtering and smoothing code at https://github.com/probml/JSL/blob/main/jsl/lds/kalman_filter.py.

**Comment by murphyk** [comment]
Here's an illustration from book 2 on how smoothing is better than filtering in the linear-Gaussian case.
We should get something qualitatively similar even if the observation model is Student not Gaussian.

Screen Shot 2022-04-16 at 11 08 14 AM

Soure code : https://github.com/probml/JSL/blob/main/jsl/demos/kf_tracking.py

**Comment by murphyk** [comment]
It would also be very cool if the heatmap illustrated true multi-modality in the filtering distribution, which then gets 'resolved' in the smoothing distribution. This might just be a question of choosing the 'right' RNG key or increasing the heavy-tailedness of the Student observation model, to make the state estimation problem more ambiguous.

**Comment by benlau6** [comment]
The bug is from `x_pdf` function and it is fixed now. Besides, `rng_key` of the particle filter example and students' t random walk example are changed to illustrate multi-modality being resolved by smoothing distribution. Line plot and density plot from student' t random walk example are attached for your reference.

Unknown-4
Unknown-3

**Comment by murphyk** [comment]
LGTM, thanks
